File: softeer/safe_robber.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int main(int argc, char** argv)
{	
	int w, n;
	cin >> w >> n;
	vector<pair<int, int>> price;
	for(int i = 0; i <n; i++){
		int m,p;
		cin >> m >> p;
		price.push_back({p,m});
	}

	sort(price.rbegin(), price.rend());

	int residual = w;
	int ans = 0;
	for(int i = 0; i < n; i ++){
		int p = price[i].first;
		int m = price[i].second;
		if(residual <= m){
			ans = ans + residual*p;
			break;
		}
		ans = ans + m*p;
		residual = residual - m;
	}

	cout << ans << endl;

	return 0;
}

File: softeer/safe_robber.py
import sys
read = sys.stdin.readline

def solution(w,n,price):

    ans = 0
    price.sort(key = lambda x : (-x[1]))

    residual = w
    for i in range(n):
        if residual <= price[i][0]:
            ans += residual * price[i][1]
            break
        residual -= price[i][0]
        ans += price[i][0] * price[i][1]

    return ans


if __name__ == "__main__":
    w,n = map(int, read().split())
    price = [list(map(int, read().split())) for _ in range(n)]

    print(solution(w,n,price))


